src/nfloat/nfixed.c

```diff
@@ -1408,3 +1456,172 @@ _nfixed_mat_mul(nn_ptr C, nn_srcptr A, nn_srcptr B, slong m, slong n, slong p, s
     else
         _nfixed_mat_mul_classical(C, A, B, m, n, p, nlimbs);
 }
+
+/*
+  Given an m x n x p matrix multiplication with inputs bounded
+  entrywise by A, B and nlimbs precision:
+
+  - Set *bound* to a bound for the entries in all intermediate
+    variables of the computation. This should be < 1 to
+    ensure correctness.
+  - Set *error* to a bound for the output error, measured in ulp.
+
+   The caller can assume that the bound is a nondecreasing function
+   of A and B.
+
+   IMPORTANT: when changing the algorithm in _nfixed_mat_mul, this
+   must be changed to correspond.
+*/
+
+void
+_nfixed_mat_mul_bound_classical(double * bound, double * error, slong m, slong n, slong p, double A, double B, slong nlimbs)
+{
+    double fixup = 1.0 + 1e-6;
```

Comment:
Why this epsilon?

Reply:
The bounds are represented as ``double`` and the arithmetic operations on them might round down, so we multiply by some factor sure to compensate for the rounding errors (in general we would also need some additive fixup like ``1e-100`` to compensate for possible underflow, but as these functions are currently used this is not an issue).